x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/changepoint/ChangePointAggregator.java

```diff
@@ -0,0 +1,391 @@
+        for (int i = 0; i < candidateChangePoints[0]; i++) {
+            lowerLeastSquares.add(i, timeWindow[i], timeWindowWeights[i]);
+        }
+        for (int i = candidateChangePoints[0], x = 0; i < timeWindow.length; i++, x++) {
+            upperLeastSquares.add(x, timeWindow[i], timeWindowWeights[i]);
+        }
+        int upperMovingWindow = 0;
+        for (int cp : candidateChangePoints) {
+            double lowerRangeVar = lowerRange.variance();
+            double upperRangeVar = upperRange.variance();
+            double rv1 = lowerLeastSquares.squareResidual();
+            double rv2 = upperLeastSquares.squareResidual();
+            double v1 = lowerRangeVar * (1 - Math.abs(rv1));
+            double v2 = upperRangeVar * (1 - Math.abs(rv2));
+            VarianceAndRValue varianceAndRValue = new VarianceAndRValue((cp * v1 + (n - cp) * v2) / n, (cp * rv1 + (n - cp) * rv2) / n);
+            if (varianceAndRValue.compareTo(vAndR) < 0) {
+                vAndR = varianceAndRValue;
+                changePoint = cp;
+            }
+            for (int i = 0; i < step; i++) {
+                lowerRange.addValue(timeWindow[i + cp], timeWindowWeights[i + cp]);
+                upperRange.removeValue(timeWindow[i + cp], timeWindowWeights[i + cp]);
+                lowerLeastSquares.add(i + cp, timeWindow[i + cp], timeWindowWeights[i + cp]);
+                upperLeastSquares.remove(i + upperMovingWindow, timeWindow[i + cp], timeWindowWeights[i + cp]);
+                upperMovingWindow++;
+            }
+        }
+        discoveredChangePoints.add(changePoint);
+
+        dfAlt = n - 6;
+        pValueVsNull = independentTrialsPValue(fTestPValue(vNull, dfNull, vAndR.variance, dfAlt), candidateChangePoints.length);
+        if (pValueVsNull < pValueThreshold && Math.abs(vAndR.rValue) >= 0.5) {
+            double pValueVsStationary = independentTrialsPValue(
+                fTestPValue(totalVariance, n - 1, vAndR.variance, dfAlt),
+                candidateChangePoints.length
+            );
+            changeType = new ChangeType.TrendChange(pValueVsStationary, vAndR.rValue, bucketValues.getBucketIndex(changePoint));
+        }
+
+        if (changeType.pValue() > 1e-5) {
+            double diff = 0.0;
+            changePoint = -1;
+            lowerRange = new RunningStats();
+            upperRange = new RunningStats();
+            // Initialize running stats so that they are only missing the individual changepoint values
+            upperRange.addValues(timeWindow, i -> timeWindowWeights[i], candidateChangePoints[0], timeWindow.length);
+            lowerRange.addValues(timeWindow, i -> timeWindowWeights[i], 0, candidateChangePoints[0]);
+            for (int cp : candidateChangePoints) {
+                double otherDiff = (0.9 * Math.abs(lowerRange.mean() - upperRange.mean())) + 0.1 * Math.abs(
+                    lowerRange.std() - upperRange.std()
+                );
```

Comment:
The distribution difference test is more sensitive when the two sets are similar size:
```suggestion
                double otherDiff = Math.min(cp, timeWindow.length - cp) * (0.9 * Math.abs(lowerRange.mean() - upperRange.mean())) + 0.1 * Math.abs(
                    lowerRange.std() - upperRange.std()
                );
```